Gemfile

```diff
@@ -33,3 +33,5 @@ end
 
 gem "tzinfo-data", platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 gem "bootstrap-sass", "3.3.7"
+gem "bcrypt"
```

Comment:
gem này để làm gì vậy?

Reply:
Gem này cung cấp hàm băm bcrypt được sử dụng bởi method has_secure_password ạ.

Reply:
Em trả lời 3 câu hỏi nhá:
- gem này được dùng để làm gì?
- gem cung cấp những method gì nổi bật?
- cơ chế hoạt động?

Reply:
Em sorry ạ cho em trả lời lại:

`bcrypt` nói chung là một dạng hàm hoặc thuật toán (hashing function hoặc cũng là một dạng key-derivation function). Nó là một hàm băm 1 chiều (tức không thể khôi phục được chuỗi ban đầu từ chuỗi đã băm).
Nó khác các hàm băm khác như MD5 hay SHA ở chỗ nó được thiết kế để chậm (giúp chống được brute force tốt hơn), nên thường được dùng để băm password.
Gem `bcrypt` giúp đem hàm băm này lên Ruby.

Cách sử dụng đơn giản là ta `require 'bcrypt'`, sau đó có thể các method sau:
Đầu tiên sử dụng 
````
my_password = BCrypt::Password.create("my password")
````
nó sẽ trả về hàm băm của chuỗi `my password`.
Nếu đã có chuỗi đã băm thì dùng như thế này: 
````
my_password = BCrypt::Password.new("$2a$12$K0ByB.6YI2/OYrB4fQOYLe6Tv0datUVf6VZ/2Jzwm879BW5K1cHey")
````
Sau đó dùng `my_password == "my password"` để kiểm tra mật khẩu xem có khớp không (trả về true hoặc false.

Cơ chế hoạt động:
`bcrypt` sẽ lấy đầu vào là một chuỗi (ví dụ như password người dùng) và sau đó tạo ra đoạn chuỗi đặc trưng của password đó. Không có cách nào lấy lại password ban đầu từ đoạn chuỗi đã băm vì quá trình này là quá trình 1 chiều.
Nhưng thường để tránh rainbow table attack, người ta sẽ prepend hoặc append một chuỗi "salt" cùng với chuỗi đầu vào nữa rồi mới chạy qua hàm băm. Salt là một chuỗi phải đảm bảo ngẫu nhiên và độc nhất.
Ngoài ra, vì máy tính càng ngày càng nhanh hơn, nên để đảm bảo `bcrypt` đủ chậm theo thời gian thì người ta còn cung cấp thêm thiết đặt `cost`, càng cao sẽ càng làm quá trình băm tốn kém hơn. Mặc định của gem `bcrypt là 12. Có thể chỉnh lại ví dụ như:
````
BCrypt::Engine.cost = 8 # Chỉnh cost mặc định
````
hoặc
````
BCrypt::Password.create('secret', :cost => 8)
````